File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {

    if(head == null){
        return null;
    }

    let slow = head, fast = head;

    for(let i = 0; i<n; i++){
        fast = fast.next;
    }
    
    if(fast == null){
        return head.next;
    }

    while(fast.next != null){
        slow = slow.next;
        fast = fast.next;
    }

    slow.next = slow.next.next;

    return head;
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private int length(ListNode head){  // O(n) n is number of nodes
        ListNode cur = head;
        int cnt = 0;
        while(cur != null){
            cnt++;
            cur = cur.next;
        }
        return cnt;
    }

    private ListNode DeleteNode(ListNode head,int pos){ // tc : O(n) & I am using recursion so sc: O(n) 
        if(head == null && pos > 0){
            return null;
        }
        if(pos == 0){
            return head.next;
        }
        else{
            head.next = DeleteNode(head.next,pos-1);
            return head;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) { // O(n)
        int size = length(head);
        return DeleteNode(head,size-n);
    }
}